In [1]:
import pandas as pd

In [2]:
braktooth_df=pd.read_excel("..//BrakTooth/Dataset.xlsx")

In [3]:
braktooth_df.head()

Protocol                                               Info  Length  \
0  HCI_EVT                              Rcvd Link Key Request       9   
1  HCI_CMD               Sent Link Key Request Negative Reply      10   
2  HCI_EVT                              Rcvd Link Key Request       9   
3  HCI_EVT  Rcvd Command Complete (Link Key Request Negati...      13   
4  HCI_CMD               Sent Link Key Request Negative Reply      10   

      Delta              Type  
0  0.010473  au_rand_flooding  
1  0.010473  au_rand_flooding  
2  0.017913  au_rand_flooding  
3  0.002642  au_rand_flooding  
4  0.002882  au_rand_flooding

In [4]:
braktooth_df.tail()

Protocol                              Info  Length     Delta    Type
8997   RFCOMM          Sent UIH Channel=12 UID       14  0.006531  normal
8998  HCI_EVT  Rcvd Number of Completed Packets       8  0.004969  normal
8999     OBEX                     Sent Continue      21  0.014489  normal
9000      LMP             LMP_power_control_req      64  0.003510  normal
9001      LMP             LMP_power_control_res      64  0.001288  normal

In [5]:
braktooth_df.shape

(9002, 5)

In [6]:
braktooth_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9002 entries, 0 to 9001
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Protocol  9002 non-null   object 
 1   Info      9002 non-null   object 
 2   Length    9002 non-null   int64  
 3   Delta     9002 non-null   float64
 4   Type      9002 non-null   object 
dtypes: float64(1), int64(1), object(3)
memory usage: 351.8+ KB


In [7]:
braktooth_df['Type'].unique()

array(['au_rand_flooding', 'duplicated_encapsulated_payload',
       'duplicated_iocap', 'truncated_sco_link_request',
       'feature_response_flooding', 'invalid_feature_page_execution',
       'lmp_auto_rate_overflow', 'lmp_overflow_dm1',
       'truncated_lmp_accepted', 'invalid_setup_complete',
       'invalid_timing_accuracy', 'normal'], dtype=object)

In [8]:
braktooth_df[braktooth_df.duplicated()]

Protocol                   Info  Length     Delta              Type
78    HCI_EVT  Rcvd Link Key Request       9  0.001042  au_rand_flooding
84    HCI_EVT  Rcvd Link Key Request       9  0.000538  au_rand_flooding
95    HCI_EVT  Rcvd Link Key Request       9  0.000571  au_rand_flooding
96    HCI_EVT  Rcvd Link Key Request       9  0.000577  au_rand_flooding
102   HCI_EVT  Rcvd Link Key Request       9  0.000565  au_rand_flooding
...       ...                    ...     ...       ...               ...
8985     OBEX     Rcvd OBEX fragment    1004  0.001877            normal
8988     OBEX     Rcvd OBEX fragment    1004  0.002029            normal
8992     OBEX     Rcvd OBEX fragment    1004  0.007267            normal
8993     OBEX     Rcvd OBEX fragment    1004  0.001653            normal
8994     OBEX     Rcvd OBEX fragment    1004  0.001159            normal

[1909 rows x 5 columns]

Observations from the dataset: -
1. The dataset has 9002 rows and 5 columns.
2. The target feature is : Type, having 12 unique values.
3. There are no null values in the dataset.
4. There are 1909 duplicate records in the dataset. Thus, we have 21.20% records as duplicates in the dataset.